README.md
# qgis-plugins
Repo for QGIS plugins


folder_creator/releases.md
## Releases

### v1.0

- Folder creation

### v.1.1

- Added support for creating grid and grid reduced Geopackages


folder_creator/test/test_resources.py
# coding=utf-8
"""Resources test.

.. note:: This program is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published by
     the Free Software Foundation; either version 2 of the License, or
     (at your option) any later version.

"""

__author__ = 'ejolaiya@integration.org'
__date__ = '2022-04-29'
__copyright__ = 'Copyright 2022, Integration Environment & Energy'

import unittest

from qgis.PyQt.QtGui import QIcon



class FolderCreatorDialogTest(unittest.TestCase):
    """Test rerources work."""

    def setUp(self):
        """Runs before each test."""
        pass

    def tearDown(self):
        """Runs after each test."""
        pass

    def test_icon_png(self):
        """Test we can click OK."""
        path = ':/plugins/FolderCreator/icon.png'
        icon = QIcon(path)
        self.assertFalse(icon.isNull())

if __name__ == "__main__":
    suite = unittest.makeSuite(FolderCreatorResourcesTest)
    runner = unittest.TextTestRunner(verbosity=2)
    runner.run(suite)





folder_creator/test/test_folder_creator_dialog.py
# coding=utf-8
"""Dialog test.

.. note:: This program is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published by
     the Free Software Foundation; either version 2 of the License, or
     (at your option) any later version.

"""

__author__ = 'ejolaiya@integration.org'
__date__ = '2022-04-29'
__copyright__ = 'Copyright 2022, Integration Environment & Energy'

import unittest

from qgis.PyQt.QtGui import QDialogButtonBox, QDialog

from folder_creator_dialog import FolderCreatorDialog

from utilities import get_qgis_app
QGIS_APP = get_qgis_app()


class FolderCreatorDialogTest(unittest.TestCase):
    """Test dialog works."""

    def setUp(self):
        """Runs before each test."""
        self.dialog = FolderCreatorDialog(None)

    def tearDown(self):
        """Runs after each test."""
        self.dialog = None

    def test_dialog_ok(self):
        """Test we can click OK."""

        button = self.dialog.button_box.button(QDialogButtonBox.Ok)
        button.click()
        result = self.dialog.result()
        self.assertEqual(result, QDialog.Accepted)

    def test_dialog_cancel(self):
        """Test we can click cancel."""
        button = self.dialog.button_box.button(QDialogButtonBox.Cancel)
        button.click()
        result = self.dialog.result()
        self.assertEqual(result, QDialog.Rejected)

if __name__ == "__main__":
    suite = unittest.makeSuite(FolderCreatorDialogTest)
    runner = unittest.TextTestRunner(verbosity=2)
    runner.run(suite)



folder_creator/resources.py
# -*- coding: utf-8 -*-

# Resource object code
#
# Created by: The Resource Compiler for PyQt5 (Qt v5.15.2)
#
# WARNING! All changes made in this file will be lost!

from PyQt5 import QtCore

qt_resource_data = b"\
\x00\x00\x04\x0a\
\x89\
\x50\x4e\x47\x0d\x0a\x1a\x0a\x00\x00\x00\x0d\x49\x48\x44\x52\x00\
\x00\x00\x17\x00\x00\x00\x18\x08\x06\x00\x00\x00\x11\x7c\x66\x75\
\x00\x00\x00\x01\x73\x52\x47\x42\x00\xae\xce\x1c\xe9\x00\x00\x00\
\x06\x62\x4b\x47\x44\x00\xff\x00\xff\x00\xff\xa0\xbd\xa7\x93\x00\
\x00\x00\x09\x70\x48\x59\x73\x00\x00\x0b\x13\x00\x00\x0b\x13\x01\
\x00\x9a\x9c\x18\x00\x00\x00\x07\x74\x49\x4d\x45\x07\xd9\x02\x15\
\x16\x11\x2c\x9d\x48\x83\xbb\x00\x00\x03\x8a\x49\x44\x41\x54\x48\
\xc7\xad\x95\x4b\x68\x5c\x55\x18\xc7\x7f\xe7\xdc\x7b\x67\xe6\xce\
\x4c\x66\x26\x49\xd3\x24\x26\xa6\xc6\xf8\x40\x21\xa5\x04\xb3\x28\
\xda\x98\x20\xa5\x0b\xad\x55\xa8\x2b\xc5\x50\x1f\xa0\x6e\x34\x2b\
\x45\x30\x14\x02\xba\x52\x69\x15\x17\x66\x63\x45\x97\x95\xa0\xad\
\x0b\xfb\xc0\x06\x25\xb6\x71\x61\x12\x41\x50\xdb\x2a\x21\xd1\xe2\
\x24\xf3\x9e\xc9\xcc\xbd\xe7\x1c\x17\x35\x43\x1e\x33\x21\xb6\xfd\
\x56\x87\xf3\x9d\xfb\xfb\x1e\xf7\xff\x9d\x23\x8c\x31\x43\x95\xf4\
\x85\x1e\x3f\x3b\x35\xac\xfd\xcc\x43\xdc\xa4\x49\x3b\xfe\x9d\x1d\
\xdb\x7b\x22\x90\x78\xf8\xb2\x28\xa7\xbe\x7d\xc1\x4b\x9d\x79\xdf\
\x18\x15\xe5\x16\x99\x10\x56\xde\x69\xdc\x3f\x22\xfd\xec\xd4\xf0\
\xad\x04\x03\x18\xa3\xa2\x7e\x76\x6a\x58\xde\x68\x2b\xb4\x36\xf8\
\xbe\xc6\x18\x53\xdb\xef\xe7\xfa\xec\xed\x67\x63\x10\x42\x00\xf0\
\xfb\xd5\x65\x2a\x15\x45\xc7\x6d\x0d\x00\xc4\xa2\xc1\xaa\x6f\x0d\
\x3e\x6c\xab\xc2\x1c\x56\xa4\x77\x4b\xb0\xf2\x35\x15\x5f\x21\x85\
\xe0\xc8\x6b\x5f\x92\x2d\x37\x33\x39\xf9\x03\x27\x8e\x1f\xa2\xf7\
\xbe\x9d\x04\x1c\x0b\x37\xe4\xac\xff\xa6\x30\x87\xbd\xba\x00\x6a\
\x06\x79\xe5\xf5\xaf\x89\xd9\x92\xc5\xcc\x0a\xd9\x7c\x19\xcf\xe9\
\xe2\xe4\xa9\x2f\x78\x7c\xff\x01\x72\x85\x0a\x2b\x65\x1f\xa5\x4c\
\xb5\xb2\x55\x16\x80\xbd\x31\xda\xda\x20\x1f\x7d\x3e\xcd\xc2\xfd\
\x59\xa6\x93\x39\x92\xd1\x22\xea\x9b\x16\xce\x9d\x3f\xce\xe0\x83\
\x03\x24\x82\x59\x3a\xdb\x7b\x88\xc7\x82\x68\x63\x58\xc9\xcc\x62\
\x8c\x21\x18\xb0\x6a\xc3\x37\x06\x49\x16\xff\x24\x6b\xa5\x49\xbb\
\x25\xbc\xa2\xa6\x21\xbb\x40\x7f\xdf\x00\x83\xbd\x01\x8e\x3c\xd5\
\x45\xd7\x8e\x6b\x9c\x9c\x98\x25\x1a\xb6\xe8\xbe\x3d\xc2\xdd\x77\
\x44\x48\xc4\x1c\x22\xe1\xeb\x58\x59\xaf\xcf\xd3\x33\x29\x2e\x34\
\x2d\x91\x93\x3e\xbe\x34\x78\x01\xc5\xe2\x61\xc5\xae\x72\x8e\x70\
\xc8\xc2\x0d\x5a\xbc\xf5\xee\x2f\x9c\xfa\x3e\x86\x69\x7a\x8e\xcf\
\x26\xe6\xf9\x63\xa1\x44\xa1\xa4\xd0\xda\x6c\x0d\x2f\x15\x7c\xb4\
\x67\x28\x59\x0a\xcf\xd6\x54\xe2\x06\x13\x87\x2b\x6f\x68\xa6\x27\
\xaf\x31\x32\x36\xc7\xb2\x7f\x17\xef\x7d\x7c\x8c\x33\x67\xcf\x12\
\x70\x24\x4a\x69\xd6\x6a\x46\xd6\xd3\x70\x72\xa9\x82\x67\x34\x45\
\xad\x28\xdb\x1a\x15\x34\x98\xff\x46\xed\xef\x37\x0d\x99\xbf\x4a\
\x3c\x30\x38\xc0\xc8\x4b\xaf\x92\x5a\x9c\xe2\xe0\x23\x6d\x74\xb4\
\xba\x84\x5d\x0b\x29\x45\x7d\xb8\x94\x82\x96\xb6\x10\xf3\xc5\x12\
\x2a\xef\x53\x11\x1a\x63\xad\x3f\x93\x19\x85\xf1\xb1\x77\x58\x5a\
\xf8\x99\x97\x9f\xe9\xa6\x75\x47\x90\xc6\xb8\x43\xd8\xb5\xb6\xce\
\xfc\xfa\xfd\x00\xfb\x3e\xf4\xc8\x05\x35\xba\x5e\xeb\x46\x21\xf9\
\xcf\x0a\xa9\x8c\x87\xe3\x48\xdc\x90\xb5\x6e\x98\x6a\xaa\x65\xf2\
\x52\x92\x43\x2f\x5e\xc2\x8c\x02\x1a\x10\xf5\x07\xac\xc3\x75\x70\
\x83\x92\x80\xb3\xf9\xd0\x26\xf8\x8f\xb3\x29\xc6\x3e\xb8\x8c\x19\
\x35\x75\x6b\x7b\x7e\x3c\xca\x45\x0c\x7e\x49\x31\xf4\x58\x3b\xf7\
\xf6\x34\x90\x88\x39\x04\x1c\x59\x1f\xfe\xdb\xd5\x3c\x5f\x9d\x4b\
\x32\xfd\x44\xb2\xba\xd7\xfa\xb6\x60\xcf\xde\x16\xdc\x90\x45\x4c\
\x4a\x2a\x9e\x62\xfe\x4e\xc5\xc8\xc1\x4e\xda\x76\x86\xe8\xe9\x0a\
\xe3\xd8\x92\x58\xd4\xc6\xb2\x44\x6d\x78\x2a\x53\xe1\xca\x7c\x99\
\x63\x5d\xbf\x56\x9d\xbd\x9f\x44\x18\x7a\xba\x95\x27\x0f\xb4\xd3\
\xdc\x18\xc0\xf3\x0d\x52\x40\xd8\xb5\xb0\xa4\x20\x14\xb2\x70\x6c\
\x81\x63\xcb\xaa\x42\xd6\xfd\xb7\xf4\xec\xa3\x06\xa0\x50\x52\xd8\
\x4e\x1b\x7e\x4a\xd3\x31\xf9\x29\xcf\xfe\xd4\x49\x7f\x5f\x13\xfb\
\xfa\x9b\x71\x43\x92\x58\xd4\x21\x18\x90\xac\xde\xb0\x42\x50\x13\
\x58\x33\xf3\x88\x6b\xa1\xfd\x65\x96\xf2\x79\xc6\x43\x7b\xd8\x75\
\x38\xcc\x3d\xdd\xd1\xaa\xcf\x71\xe4\xff\x7f\x91\x56\x33\xaf\xea\
\x37\xe7\xa1\x94\x21\x16\xb5\xd1\x06\x2c\x29\x36\xf5\x72\x9b\x96\
\x95\xc0\xc4\xda\x9d\x78\x83\x43\x53\x22\x80\x65\x09\x1c\xfb\x86\
\xc1\x00\xe7\x25\x70\x14\x48\x6f\x1e\x22\x51\xe3\x75\xd9\xb6\xa5\
\x81\xa3\x32\xb1\xfb\xf4\x0c\x30\xb8\xb1\x82\x9b\xb0\x09\x60\x30\
\xb1\xfb\xf4\xcc\xbf\xa0\xe9\x6e\xae\x5a\xdf\x4b\x81\x00\x00\x00\
\x00\x49\x45\x4e\x44\xae\x42\x60\x82\
"

qt_resource_name = b"\
\x00\x07\
\x07\x3b\xe0\xb3\
\x00\x70\
\x00\x6c\x00\x75\x00\x67\x00\x69\x00\x6e\x00\x73\
\x00\x0e\
\x06\xde\x10\x82\
\x00\x66\
\x00\x6f\x00\x6c\x00\x64\x00\x65\x00\x72\x00\x5f\x00\x63\x00\x72\x00\x65\x00\x61\x00\x74\x00\x6f\x00\x72\
\x00\x08\
\x0a\x61\x5a\xa7\
\x00\x69\
\x00\x63\x00\x6f\x00\x6e\x00\x2e\x00\x70\x00\x6e\x00\x67\
"

qt_resource_struct_v1 = b"\
\x00\x00\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x01\
\x00\x00\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x02\
\x00\x00\x00\x14\x00\x02\x00\x00\x00\x01\x00\x00\x00\x03\
\x00\x00\x00\x36\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\
"

qt_resource_struct_v2 = b"\
\x00\x00\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x01\
\x00\x00\x00\x00\x00\x00\x00\x00\
\x00\x00\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x02\
\x00\x00\x00\x00\x00\x00\x00\x00\
\x00\x00\x00\x14\x00\x02\x00\x00\x00\x01\x00\x00\x00\x03\
\x00\x00\x00\x00\x00\x00\x00\x00\
\x00\x00\x00\x36\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\
\x00\x00\x01\x80\x75\x80\x82\x76\
"

qt_version = [int(v) for v in QtCore.qVersion().split('.')]
if qt_version < [5, 8, 0]:
    rcc_version = 1
    qt_resource_struct = qt_resource_struct_v1
else:
    rcc_version = 2
    qt_resource_struct = qt_resource_struct_v2

def qInitResources():
    QtCore.qRegisterResourceData(rcc_version, qt_resource_struct, qt_resource_name, qt_resource_data)

def qCleanupResources():
    QtCore.qUnregisterResourceData(rcc_version, qt_resource_struct, qt_resource_name, qt_resource_data)

qInitResources()


folder_creator/folder_creator_dialog.py
# -*- coding: utf-8 -*-
"""
/***************************************************************************
 FolderCreatorDialog
                                 A QGIS plugin
 This plugin automates the creation of folders for each settlements in a work order
 Generated by Plugin Builder: http://g-sherman.github.io/Qgis-Plugin-Builder/
                             -------------------
        begin                : 2022-04-29
        git sha              : $Format:%H$
        copyright            : (C) 2022 by Integration Environment & Energy
        email                : ejolaiya@integration.org
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
"""

import os

from qgis.PyQt import uic
from qgis.PyQt import QtWidgets

# This loads your .ui file so that PyQt can populate your plugin with the elements from Qt Designer
FORM_CLASS, _ = uic.loadUiType(os.path.join(
    os.path.dirname(__file__), 'folder_creator_dialog_base.ui'))


class FolderCreatorDialog(QtWidgets.QDialog, FORM_CLASS):
    def __init__(self, parent=None):
        """Constructor."""
        super(FolderCreatorDialog, self).__init__(parent)
        # Set up the user interface from Designer through FORM_CLASS.
        # After self.setupUi() you can access any designer object by doing
        # self.<objectname>, and you can use autoconnect slots - see
        # http://qt-project.org/doc/qt-4.8/designer-using-a-ui-file.html
        # #widgets-and-dialogs-with-auto-connect
        self.setupUi(self)


folder_creator/models_and_styles/grid.py
"""
Model exported as python.
Name : A4 field paper
Group : field paper
With QGIS : 32003
"""

from qgis.core import QgsProcessing
from qgis.core import QgsProcessingAlgorithm
from qgis.core import QgsProcessingMultiStepFeedback
from qgis.core import QgsProcessingParameterVectorLayer
from qgis.core import QgsProcessingParameterFeatureSink
from qgis.core import QgsCoordinateReferenceSystem
from qgis.core import QgsExpression
import processing
import os

class A4FieldPaper(QgsProcessingAlgorithm):

    def initAlgorithm(self, config=None):
        self.addParameter(QgsProcessingParameterVectorLayer('residentialboundary', 'Residential boundary', types=[QgsProcessing.TypeVectorPolygon], defaultValue=None))
        self.addParameter(QgsProcessingParameterFeatureSink('Final', 'Final', type=QgsProcessing.TypeVectorAnyGeometry, createByDefault=True, supportsAppend=True, defaultValue=None))

    def processAlgorithm(self, parameters, context, model_feedback):
        # Use a multi-step feedback, so that individual child algorithm progress reports are adjusted for the
        # overall progress through the model
        feedback = QgsProcessingMultiStepFeedback(6, model_feedback)
        results = {}
        outputs = {}

        # Create grid
        alg_params = {
            'CRS': QgsCoordinateReferenceSystem('EPSG:3857'),
            'EXTENT': parameters['residentialboundary'],
            'HOVERLAY': 0,
            'HSPACING': 195,
            'TYPE': 2,  # Rectangle (Polygon)
            'VOVERLAY': 0,
            'VSPACING': 235,
            'OUTPUT': QgsProcessing.TEMPORARY_OUTPUT
        }
        outputs['CreateGrid'] = processing.run('native:creategrid', alg_params, context=context, feedback=feedback, is_child_algorithm=True)

        feedback.setCurrentStep(1)
        if feedback.isCanceled():
            return {}

        # COL Field calculator
        alg_params = {
            'FIELD_LENGTH': 50,
            'FIELD_NAME': 'col',
            'FIELD_PRECISION': 0,
            'FIELD_TYPE': 0,  # Float
            'FORMULA': QgsExpression('\'((\' || \'left\' || \' - \' ||  \'minimum(left)\'  || \')/195)+1\'').evaluate(),
            'INPUT': outputs['CreateGrid']['OUTPUT'],
            'OUTPUT': QgsProcessing.TEMPORARY_OUTPUT
        }
        
        outputs['ColFieldCalculator'] = processing.run('native:fieldcalculator', alg_params, context=context, feedback=feedback, is_child_algorithm=True)
        
        feedback.setCurrentStep(2)
        if feedback.isCanceled():
            return {}

        # ROW Field calculator
        alg_params = {
            'FIELD_LENGTH': 50,
            'FIELD_NAME': 'row',
            'FIELD_PRECISION': 3,
            'FIELD_TYPE': 0,  # Float
            'FORMULA': QgsExpression('\'((\' || \'top\' || \' - \' ||  \'maximum(top)\'  || \')/235)*-1\'').evaluate(),
            'INPUT': outputs['ColFieldCalculator']['OUTPUT'],
            'OUTPUT': QgsProcessing.TEMPORARY_OUTPUT
        }
        outputs['RowFieldCalculator'] = processing.run('native:fieldcalculator', alg_params, context=context, feedback=feedback, is_child_algorithm=True)

        feedback.setCurrentStep(3)
        if feedback.isCanceled():
            return {}

        # Letter Field calculator
        alg_params = {
            'FIELD_LENGTH': 6,
            'FIELD_NAME': 'let',
            'FIELD_PRECISION': 0,
            'FIELD_TYPE': 2,  # String
            'FORMULA': QgsExpression(' \'char(65+\' || \'row\' || \')\'').evaluate(),
            'INPUT': outputs['RowFieldCalculator']['OUTPUT'],
            'OUTPUT': QgsProcessing.TEMPORARY_OUTPUT
        }
        outputs['LetterFieldCalculator'] = processing.run('native:fieldcalculator', alg_params, context=context, feedback=feedback, is_child_algorithm=True)

        feedback.setCurrentStep(4)
        if feedback.isCanceled():
            return {}

        # Final field calculator
        alg_params = {
            'FIELD_LENGTH': 5,
            'FIELD_NAME': 'g_id',
            'FIELD_PRECISION': 0,
            'FIELD_TYPE': 2,  # String
            'FORMULA': QgsExpression(' \'concat(\'|| \'let\' ||\',\'|| \'col\' || \')\'').evaluate(),
            'INPUT': outputs['LetterFieldCalculator']['OUTPUT'],
            'OUTPUT': parameters['Final']
        }
        outputs['FinalFieldCalculator'] = processing.run('native:fieldcalculator', alg_params, context=context, feedback=feedback, is_child_algorithm=True)
        results['Final'] = outputs['FinalFieldCalculator']['OUTPUT']

        feedback.setCurrentStep(5)
        if feedback.isCanceled():
            return {}

        # Set style for vector layer
        basepath = os.path.dirname(__file__)
        style_path = os.path.join(basepath,'styles/a4_grid_style.qml')
        alg_params = {
            'INPUT': outputs['FinalFieldCalculator']['OUTPUT'],
            'STYLE': style_path
        }
        outputs['SetStyleForVectorLayer'] = processing.run('qgis:setstyleforvectorlayer', alg_params, context=context, feedback=feedback, is_child_algorithm=True)
        
        
        return results

    def name(self):
        return 'A4 field paper'

    def displayName(self):
        return 'A4 field paper'

    def group(self):
        return 'field paper'

    def groupId(self):
        return 'field paper'

    def createInstance(self):
        return A4FieldPaper()


folder_creator/models_and_styles/grid_reduced.py
"""
Model exported as python.
Name : grid_reduced
Group : 
With QGIS : 32003
"""

from qgis.core import QgsProcessing
from qgis.core import QgsProcessingAlgorithm
from qgis.core import QgsProcessingMultiStepFeedback
from qgis.core import QgsProcessingParameterVectorLayer
from qgis.core import QgsProcessingParameterFeatureSink
import processing
import os

class Grid_reduced(QgsProcessingAlgorithm):

    def initAlgorithm(self, config=None):
        self.addParameter(QgsProcessingParameterVectorLayer('grid', 'grid', types=[QgsProcessing.TypeVectorPolygon], defaultValue=None))
        self.addParameter(QgsProcessingParameterVectorLayer('outline', 'outline', types=[QgsProcessing.TypeVectorPolygon], defaultValue=None))
        self.addParameter(QgsProcessingParameterFeatureSink('Grid_reduced', 'grid_reduced', type=QgsProcessing.TypeVectorAnyGeometry, createByDefault=True, defaultValue=None))

    def processAlgorithm(self, parameters, context, model_feedback):
        # Use a multi-step feedback, so that individual child algorithm progress reports are adjusted for the
        # overall progress through the model
        feedback = QgsProcessingMultiStepFeedback(2, model_feedback)
        results = {}
        outputs = {}

        # Extract by location
        alg_params = {
            'INPUT': parameters['grid'],
            'INTERSECT': parameters['outline'],
            'PREDICATE': [0],  # intersect
            'OUTPUT': parameters['Grid_reduced']
        }
        outputs['ExtractByLocation'] = processing.run('native:extractbylocation', alg_params, context=context, feedback=feedback, is_child_algorithm=True)
        results['Grid_reduced'] = outputs['ExtractByLocation']['OUTPUT']

        feedback.setCurrentStep(1)
        if feedback.isCanceled():
            return {}

        basepath = os.path.dirname(__file__)
        style_path = os.path.join(basepath,'styles/a4_grid_reduced_style.qml')
        # Set style for vector layer
        alg_params = {
            'INPUT': outputs['ExtractByLocation']['OUTPUT'],
            'STYLE': style_path
        }
        outputs['SetStyleForVectorLayer'] = processing.run('qgis:setstyleforvectorlayer', alg_params, context=context, feedback=feedback, is_child_algorithm=True)
        return results

    def name(self):
        return 'grid_reduced'

    def displayName(self):
        return 'grid_reduced'

    def group(self):
        return ''

    def groupId(self):
        return ''

    def createInstance(self):
        return Grid_reduced()


folder_creator/folder_creator.py
# -*- coding: utf-8 -*-
"""
/***************************************************************************
 FolderCreator
                                 A QGIS plugin
 This plugin automates the creation of folders for each settlements in a work order
 Generated by Plugin Builder: http://g-sherman.github.io/Qgis-Plugin-Builder/
                              -------------------
        begin                : 2022-04-29
        git sha              : $Format:%H$
        copyright            : (C) 2022 by Integration Environment & Energy
        email                : ejolaiya@integration.org
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
"""

from math import ceil
import shutil





from qgis.PyQt.QtCore import QSettings, QTranslator, QCoreApplication
from qgis.PyQt.QtGui import QIcon
from qgis.PyQt.QtWidgets import QAction
from qgis.core import Qgis,QgsVectorLayer,QgsVectorFileWriter,QgsProject
from qgis.core import QgsProcessingFeedback,QgsProcessingContext
# Initialize Qt resources from file resources.py
from .resources import *
# Import the code for the dialog
from .folder_creator_dialog import FolderCreatorDialog
import os

# import and initialize models
from .models_and_styles import Grid_reduced,A4FieldPaper


class FolderCreator:
    """QGIS Plugin Implementation."""

    def __init__(self, iface):
        """Constructor.

        :param iface: An interface instance that will be passed to this class
            which provides the hook by which you can manipulate the QGIS
            application at run time.
        :type iface: QgsInterface
        """
        # Save reference to the QGIS interface
        self.iface = iface
        # initialize plugin directory
        self.plugin_dir = os.path.dirname(__file__)
        # initialize locale
        locale = QSettings().value('locale/userLocale')[0:2]
        locale_path = os.path.join(
            self.plugin_dir,
            'i18n',
            'FolderCreator_{}.qm'.format(locale))

        if os.path.exists(locale_path):
            self.translator = QTranslator()
            self.translator.load(locale_path)
            QCoreApplication.installTranslator(self.translator)

        # Declare instance attributes
        self.actions = []
        self.menu = self.tr(u'&Folder Creator')

        # Check if plugin was started the first time in current QGIS session
        # Must be set in initGui() to survive plugin reloads
        self.first_start = None

    def generate_grids_and_grid_reduced(self,grid_path,type,feature,cluster_path=None,grid_reduced_path=None):
        
       

        if type == "grid":
             # create layer
            vector_layer = QgsVectorLayer("Polygon", "temporary_polygon", "memory")
            data_provider = vector_layer.dataProvider()

            # add a feature
            data_provider.addFeatures([feature])
            vector_layer.updateExtents()
            grid = A4FieldPaper()
            feedback =  QgsProcessingFeedback()
            model_context = QgsProcessingContext()

            parameters = {
                    "residentialboundary":vector_layer,
                    "Final":grid_path
            }
            grid.initAlgorithm()
            grid.processAlgorithm(parameters=parameters,model_feedback=feedback,context=model_context)
            del(vector_layer)
        
        if type == "grid_reduced":

            grid_reduced = Grid_reduced()
            feedback =  QgsProcessingFeedback()
            model_context = QgsProcessingContext()

            parameters = {
                    "grid":grid_path,
                    "outline":cluster_path,
                    "Grid_reduced":grid_reduced_path
            }
            grid_reduced.initAlgorithm()
            grid_reduced.processAlgorithm(parameters=parameters,model_feedback=feedback,context=model_context)

    def generate_folders(self):

        joined_layer:QgsVectorLayer = self.dlg.joined_layer.currentLayer()
        work_order_number:int = self.dlg.work_order_number.value()
        state:str= self.dlg.state_name.text() 
        settlement_id_name:str = 'Name' if self.dlg.settlement_id_column_name.text() =="" else self.dlg.settlement_id_column_name.text()
        output_folder:str = self.dlg.output_folder.filePath()
        progress_bar = self.dlg.progress_bar
        
        #sanity checks

        if not (joined_layer.isSpatial() and isinstance(joined_layer,QgsVectorLayer)):
            
            self.iface.messageBar().pushMessage(
                "Layer error", "Selected layer is not supported. Plugin only support a vector Layer(GeoPackage, Shapefile etc)",level=Qgis.Critical,duration=5)

        elif state == "":
            self.iface.messageBar().pushMessage(
                "State error", "State name not provided",level=Qgis.Critical,duration=5
            )
        
        elif not os.path.exists(output_folder):

            self.iface.messageBar().pushMessage(
                "Path error", "Provided output folder does not exist",level=Qgis.Critical,duration=5
            )

        elif not joined_layer.hasFeatures():
            
            self.iface.messageBar().pushMessage(
                "Layer error", "Provided layer does not have a feature",level=Qgis.Critical,duration=5
            )
        
        else:


            total_features = len([x for x in joined_layer.getFeatures()])
            multiplication_factor = ceil(100/total_features)


            #QGIS output file options
            options=QgsVectorFileWriter.SaveVectorOptions()
            options.driverName="GPKG"
            options.onlySelectedFeatures = True
            context = QgsProject.instance().transformContext()

           
            for index,feature in enumerate(joined_layer.getFeatures()):
                
                joined_layer.select(feature.id())
                progress_bar.setValue(multiplication_factor * index + multiplication_factor) #needs some touches

                cluster_id = feature['cluster_of']  #assumed all cluster id column name are cluster_of
                settlement_id = feature[settlement_id_name]
                cleaned_cluster_id = str(cluster_id).split(".")[0] if isinstance(cluster_id,float) else cluster_id
                directory_name = f"{state}_WO{work_order_number}-{settlement_id}_{cleaned_cluster_id}"
                file_name = f"cluster_{cleaned_cluster_id}.gpkg"
                full_path = os.path.join(output_folder,directory_name)
                full_file_path = os.path.join(full_path,file_name)
                grid_path = os.path.join(full_path,"grid.gpkg")
                grid_reduced_path = os.path.join(full_path,"grid_reduced.gpkg")

                if not os.path.exists(full_path):
                    os.mkdir(full_path)
                    _writer = QgsVectorFileWriter.writeAsVectorFormatV3(
                    joined_layer,full_file_path,context,options
                    )

                    #generate grids and grid reduced
                    self.generate_grids_and_grid_reduced(grid_path=grid_path,type="grid",feature=feature)
                    self.generate_grids_and_grid_reduced(
                        grid_path=grid_path,
                        grid_reduced_path=grid_reduced_path,
                        cluster_path=full_file_path,
                        type="grid_reduced",
                        feature=feature
                    )
                    joined_layer.removeSelection()

                else:
                    shutil.rmtree(full_path)
                    os.mkdir(full_path)
                    _writer = QgsVectorFileWriter.writeAsVectorFormatV3(
                        joined_layer,full_file_path,context,options
                    )
                    
                    #generate grids and grid reduced
                    self.generate_grids_and_grid_reduced(grid_path=grid_path,type="grid",feature=feature)
                    self.generate_grids_and_grid_reduced(
                        grid_path=grid_path,
                        grid_reduced_path=grid_reduced_path,
                        cluster_path=full_file_path,
                        type="grid_reduced",
                        feature=feature
                    )
                    joined_layer.removeSelection()
                
            
            self.iface.messageBar().pushMessage("Success", f"Operation completed,check here for generated folders: {full_path}",level=Qgis.Success,duration=5)
            
    # noinspection PyMethodMayBeStatic
    def tr(self, message):
        """Get the translation for a string using Qt translation API.

        We implement this ourselves since we do not inherit QObject.

        :param message: String for translation.
        :type message: str, QString

        :returns: Translated version of message.
        :rtype: QString
        """
        # noinspection PyTypeChecker,PyArgumentList,PyCallByClass
        return QCoreApplication.translate('FolderCreator', message)

    def add_action(
        self,
        icon_path,
        text,
        callback,
        enabled_flag=True,
        add_to_menu=True,
        add_to_toolbar=True,
        status_tip=None,
        whats_this=None,
        parent=None):
        """Add a toolbar icon to the toolbar.

        :param icon_path: Path to the icon for this action. Can be a resource
            path (e.g. ':/plugins/foo/bar.png') or a normal file system path.
        :type icon_path: str

        :param text: Text that should be shown in menu items for this action.
        :type text: str

        :param callback: Function to be called when the action is triggered.
        :type callback: function

        :param enabled_flag: A flag indicating if the action should be enabled
            by default. Defaults to True.
        :type enabled_flag: bool

        :param add_to_menu: Flag indicating whether the action should also
            be added to the menu. Defaults to True.
        :type add_to_menu: bool

        :param add_to_toolbar: Flag indicating whether the action should also
            be added to the toolbar. Defaults to True.
        :type add_to_toolbar: bool

        :param status_tip: Optional text to show in a popup when mouse pointer
            hovers over the action.
        :type status_tip: str

        :param parent: Parent widget for the new action. Defaults None.
        :type parent: QWidget

        :param whats_this: Optional text to show in the status bar when the
            mouse pointer hovers over the action.

        :returns: The action that was created. Note that the action is also
            added to self.actions list.
        :rtype: QAction
        """

        icon = QIcon(icon_path)
        action = QAction(icon, text, parent)
        action.triggered.connect(callback)
        action.setEnabled(enabled_flag)

        if status_tip is not None:
            action.setStatusTip(status_tip)

        if whats_this is not None:
            action.setWhatsThis(whats_this)

        if add_to_toolbar:
            # Adds plugin icon to Plugins toolbar
            self.iface.addToolBarIcon(action)

        if add_to_menu:
            self.iface.addPluginToMenu(
                self.menu,
                action)

        self.actions.append(action)

        return action

    def initGui(self):
        """Create the menu entries and toolbar icons inside the QGIS GUI."""

        icon_path = ':/plugins/folder_creator/icon.png'
        self.add_action(
            icon_path,
            text=self.tr(u'Automated Folder Creation'),
            callback=self.run,
            parent=self.iface.mainWindow())

        # will be set False in run()
        self.first_start = True

    def unload(self):
        """Removes the plugin menu item and icon from QGIS GUI."""
        for action in self.actions:
            self.iface.removePluginMenu(
                self.tr(u'&Folder Creator'),
                action)
            self.iface.removeToolBarIcon(action)

    def run(self):
        """Run method that performs all the real work"""

        # Create the dialog with elements (after translation) and keep reference
        # Only create GUI ONCE in callback, so that it will only load when the plugin is started
        if self.first_start == True:
            self.first_start = False
            self.dlg = FolderCreatorDialog()
            
        self.dlg.generate_button.clicked.connect(self.generate_folders)
        # show the dialog
        self.dlg.show()
        self.dlg.progress_bar.setValue(0)
        # Run the dialog event loop
        result = self.dlg.exec_()
        
        # See if OK was pressed
        if result:
            # Do something useful here - delete the line containing pass and
            # substitute with your code.
            pass 









